composer/models/nlp_metrics.py

```diff
@@ -9,6 +9,149 @@
+
+    def update(self, preds: torch.Tensor, target: torch.Tensor):
+        # predictions is a batch x num_classes tensor, take the argmax to get class indicies
+        preds = torch.argmax(preds, dim=-1)
+        assert preds.shape == target.shape
+
+        # mask out the padded indicies
+        mask = (target != self.ignore_index)
+        masked_target = target[mask]
+        masked_preds = preds[mask]
+
+        self.correct += torch.sum(masked_preds == masked_target)
+        self.total += mask.sum()
+
+    def compute(self):
+        assert isinstance(self.correct, Tensor)
+        assert isinstance(self.total, Tensor)
+        return self.correct.float() / self.total
+
+
+# TODO (Moin): write tests for this!
+class CrossEntropyLoss(Metric):
+    """Computes cross entropy loss.
+
+    Args:
+        vocab_size (int): the size of the tokenizer vocabulary.
+        dist_sync_on_step (bool): Synchronize metric state across processes at
+            each forward() before returning the value at the step.
+        ignore_index (int): The class index to ignore. Defaults to -100.
+
+    State:
+        sum_loss (float): the sum of the per-example loss in the batch.
+        total_items (float): the number of batches to average across.
+    """
+
+    def __init__(self, vocab_size: int, dist_sync_on_step=False, ignore_index: int = -100):
+        super().__init__(dist_sync_on_step=dist_sync_on_step)
+
+        self.vocab_size = vocab_size
+        self.ignore_index = ignore_index
+        self.loss_fn = torch.nn.CrossEntropyLoss(ignore_index=ignore_index, reduction="sum")
+        self.add_state("sum_loss", default=torch.tensor(0.), dist_reduce_fx="sum")
+        self.add_state("total_items", default=torch.tensor(0), dist_reduce_fx="sum")
+
+    def update(self, output: Union[Mapping, Tensor], target: Tensor) -> None:
+        """Updates the internal state with results from a new batch.
+
+        Args:
+            output (Mapping): The output from the model, which must contain
+                either the Tensor or a Mapping type that contains the loss or model logits.
+            target (Tensor): A Tensor of ground-truth values to compare against.
+        """
+
+        assert isinstance(output, Tensor)
+        output = output.view(-1, self.vocab_size)
+        target = target.view(-1)
+        losses = self.loss_fn(output, target)
+
+        total_items = (target != self.ignore_index).sum()
+        self.total_items += total_items  #type: ignore (third-party)
+
+        # accmulate loss over all batches
+        self.sum_loss += losses
+
+    def compute(self) -> Tensor:
+        """Aggregate the state over all processes to compute the metric.
+
+        Returns:
+            loss (Tensor): The loss averaged across all batches.
+        """
+        # Return average loss over entire dataset
+        return self.sum_loss / self.total_items  #type: ignore (third-party)
+
+
+# TODO (Moin): write tests for this!
```

Comment:
Ah last thing (I think), could you remove these TODOs?

Reply:
ooh good catch! Let me actually grep for TODOs overall

Reply:
Done!